### DISCUSSION TITLE
Setting init-only indexer in `with` expression initializer list?

### DISCUSSION THREAD

**Discussion by dmchurch** [dotnet/csharplang#8117] [Q&A]
I'm having trouble figuring out if this is a feature gap or if I'm just using it wrong. I've got a bunch of small `readonly struct`s that I'm using in a performance-critical section of my code, and I'm using the `with` expression from C#10 to do nondestructive mutation. I thought that you could use any syntax you could in a `new() { ... }` initializer list, but the compiler complains if I try and assign to an indexer:

```csharp
[StructLayout(LayoutKind.Sequential, Size = 4)]
readonly struct IPv4
{
    public readonly byte A { get; init; }
    public readonly byte B { get; init; }
    public readonly byte C { get; init; }
    public readonly byte D { get; init; }
    public byte this[int octet]
    {
        get => octet switch
        {
            0 => A,
            1 => B,
            2 => C,
            3 => D,
            _ => throw new IndexOutOfRangeException()
        };
        init
        {
            switch (octet) {
                case 0: A = value; break;
                case 1: B = value; break;
                case 2: C = value; break;
                case 3: D = value; break;
                default: throw new IndexOutOfRangeException();
            }
        }
    }

    public static IPv4 ClassARoot  => new IPv4() { A = 10 };     // Valid
    public IPv4 ClassCNetwork      =>  this with { D = 0 };      // Valid
    public static IPv4 ClassARoot2 => new IPv4() { [0] = 10 };   // Valid
    public IPv4 ClassCNetwork2     =>  this with { [3] = 0 };    // Syntax error (CS0747)
}
```

Shouldn't indexers be treated the same as other properties in `with` expressions? If so, what's wrong with my code?

**Comment by HaloFour** [comment]
> Shouldn't indexers be treated the same as other properties in `with` expressions? 

Withers were not implemented to work with indexers, only properties.  I don't know if there was an explicit decision against supporting indexers, there probably wasn't a good use case for it.  I can imagine it being considered a pit of failure given that unless the type is carefully designed that such an indexer would result in mutating both the instances of the type.

**Reply by dmchurch** [reply]
> unless the type is carefully designed that such an indexer would result in mutating both the instances of the type.

Can you clarify what you mean by this? Indexers are basically just properties with parameters - not sure how that makes them more or less dangerous than other properties?

**Reply by HaloFour** [reply]
> Indexers are basically just properties with parameters - not sure how that makes them more or less dangerous than other properties?

Intent and typical usage patterns.  Properties are generally independent scalar values, that's not the case with indexers.  I would suggest that your example would be a very niche case and that indexers typically would reference some internal array or other collection, which would require additional work on the part of the developer to clone correctly to support for "withers".
